tiledb/sm/filesystem/azure.cc

```diff
@@ -0,0 +1,716 @@
+  return Status::Ok();
+}
+
+Status Azure::is_dir(const URI& uri, bool* const exists) const {
+  assert(client_);
+  assert(exists);
+
+  // Potentially add `/` to the end of `uri`
+  const URI uri_dir = uri.add_trailing_slash();
+  std::vector<std::string> paths;
+  RETURN_NOT_OK(ls(uri_dir, &paths, "/", 1));
+  *exists = (bool)paths.size();
+  return Status::Ok();
+}
+
+Status Azure::is_blob(const URI& uri, bool* const is_blob) const {
+  assert(is_blob);
+
+  std::string container_name;
+  std::string blob_path;
+  RETURN_NOT_OK(parse_azure_uri(uri, &container_name, &blob_path));
+
+  return this->is_blob(container_name, blob_path, is_blob);
+}
+
+Status Azure::is_blob(
+    const std::string& container_name,
+    const std::string& blob_path,
+    bool* const is_blob) const {
+  assert(client_);
+  assert(is_blob);
+
+  std::future<
+      azure::storage_lite::storage_outcome<azure::storage_lite::blob_property>>
+      request = client_->get_blob_properties(container_name, blob_path);
+  if (!request.valid()) {
+    return LOG_STATUS(Status::AzureError(
+        std::string("Get blob properties failed on: " + blob_path)));
+  }
+
+  azure::storage_lite::storage_outcome<azure::storage_lite::blob_property>
+      outcome = request.get();
+  if (!outcome.success()) {
+    *is_blob = false;
+    return Status::Ok();
+  }
+
+  azure::storage_lite::blob_property response = outcome.response();
+
+  *is_blob = response.valid();
+  return Status::Ok();
+}
+
+Status Azure::ls(
+    const URI& uri,
+    std::vector<std::string>* paths,
+    const std::string& delimiter,
+    const int max_paths) const {
+  assert(client_);
+  assert(paths);
+
+  if (!uri.is_azure()) {
+    return LOG_STATUS(Status::AzureError(
+        std::string("URI is not an Azure URI: " + uri.to_string())));
+  }
+
+  std::string container_name;
+  std::string blob_path;
+  RETURN_NOT_OK(parse_azure_uri(uri, &container_name, &blob_path));
+
+  std::string continuation_token = "";
+  do {
+    std::future<azure::storage_lite::storage_outcome<
+        azure::storage_lite::list_blobs_segmented_response>>
+        request = client_->list_blobs_segmented(
+            container_name,
+            delimiter,
+            continuation_token,
+            blob_path,
+            max_paths > 0 ? max_paths : 5000);
+    if (!request.valid()) {
+      return LOG_STATUS(Status::AzureError(
+          std::string("List blobs failed on: " + uri.to_string())));
+    }
+
+    azure::storage_lite::storage_outcome<
+        azure::storage_lite::list_blobs_segmented_response>
+        outcome = request.get();
+    if (!outcome.success()) {
+      return LOG_STATUS(Status::AzureError(
+          std::string("List blobs failed on: " + uri.to_string())));
+    }
+
+    azure::storage_lite::list_blobs_segmented_response response =
+        outcome.response();
+
+    for (const auto& blob : response.blobs) {
+      paths->emplace_back("azure://" + container_name + "/" + blob.name);
```

Comment:
Just for consistency, I'll keep it this way to match S3 and create a separate clubhouse ticket to provide URI normalize routines for both. Just so that they get checked in at the same time.